src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -1206,6 +1237,44 @@ private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectE
             return startInfo;
         }
 
+        /// <summary>
+        /// Determine if we have a special file which will change the way native argument passing
+        /// is done on Windows. We use legacy behavior for cmd.exe, .bat, .cmd files.
+        /// </summary>
+        /// <param name="filename">The file to use when checking how to pass arguments.</param>
+        /// <returns>A boolean indicating what passing style should be used.</returns>
+        private static bool UseLegacyPassingStyle(string filename)
+        {
+            if (filename == null || filename == string.Empty)
+            {
+                return false;
+            }
+
+            string commandPath = filename.ToLowerInvariant();
+
+            // This is the list of files which will trigger Legacy behavior if
+            // PSNativeCommandArgumentPassing is set to "Windows".
+            // The following native commands have non-standard behavior with regard to argument passing.
+            string[] exceptions = new string[]
+                {
+                "cmd.exe",
+                "cscript.exe",
+                "wscript.exe",
+                ".bat",
+                ".cmd",
+                ".vbs"
+                };
+            foreach (string exception in exceptions)
```

Comment:
Since all I have is the path as a string (not a fileinfo), I'll need to do a string comparison anyway and I can't do "HashSet.Contains (or similar). I sort of feel that creating 2 lists is less efficient than 1.

Reply:
This is a full path?  wouldn't that work incorrectly for `mycmd.exe`?

Reply:
@JamesWTruher I don't think this comment has been addressed?

Reply:
yes, it's same for cscript.exe and wscript.exe (not sure why, but it would be possible). I will add checks for explicit "\\cmd.exe" and "/cmd.exe"

Reply:
I think it would be better to get the FileInfo and have separate hashtable for filename and fileextension.  Alternatively, the exception behavior is only for Windows so we could `#ifdef` that portion and not worry about forward slashes as I would expect the filePath to be normalized by this point?

Reply:
Rather than try to hand roll this in any particular way, given that we're touching the file system and about to invoke a process, I'd just eat the allocations and do something like this:

```csharp
private static readonly IReadOnlySet<string> s_legacyFileExtensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
{
    ".js",
    ".wsf",
    ".cmd",
    ".bat",
    ".vbs",
}

private static readonly IReadOnlySet<string> s_legacyCommands = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
{
    "cmd",
    "cscript",
    "wscript",
};

private static bool ShouldUseLegacyBehaviour(string commandPath)
{
    ReadOnlySpan<char> extension = Path.GetExtension(commandPath);

    if (s_legacyFileExtensions.Contains(extension))
    {
        return true;
    }

    ReadOnlySpan<char> commandName = Path.GetFileNameWithoutExtension(commandPath);

    if (s_legacyCommands.Contains(commandName))
    {
        return true;
    }

    return false;
}
```

If we can work out how to make the HashSets work with `ReadOnlySpan<char>` we can avoid allocations anyway...

Reply:
> Alternatively, the exception behavior is only for Windows so we could #ifdef that portion and not worry about forward slashes as I would expect the filePath to be normalized by this point?

NativeCommandProcessor doesn't normalize the path. I am not sure it gets the path normalized.

> I think it would be better to get the FileInfo and have separate hashtable for filename and fileextension.

I'm afraid this array is too small for us to see the benefits of the hashtable/hashset.

Reply:
```csharp
private static bool ShouldUseLegacyParsing(string commandPath) =>
    !string.IsNullOrEmpty(commandPath)
    && (s_legacyFileExtensions.Contains(Path.GetExtension(commandPath))
        || s_legacyCommands.Contains(Path.GetFileNameWithoutExtension(commandPath)));
```

Reply:
@JamesWTruher do you have concerns with changing to Rob's proposal?  I would prefer leveraging the Path methods.

Reply:
I think what I propose here would still be the right way to go